### PR TITLE
Streamable EPT Reader

### PR DESCRIPTION
**Pull request by pravinshinde825**
In this PR I tried to make EPT Reader streamable. 
This PR is supposed to be submitted to PDAL directly. But I thought, first get it reviewed from our experts. 

**Test pipeline :**
```
{
   "pipeline": [
       {
           "type": "readers.ept",
           "filename": "gs://helix-dev-integration-test-data/pcp/1554304352"
       },
       {
           "type": "writers.las",
           "filename": "D:/ept-to-laz.laz"
       }
   ]
}
```
This works with both Binary as well as laszip ept pointcloud.

### CODE REVIEW

**Review by vaibhav-bhandari — APPROVED** [review]

File: io/EptReader.cpp [hunk 1]
```diff
@@ -613,14 +618,18 @@ PointViewSet EptReader::run(PointViewPtr view)
 }
 
 uint64_t EptReader::readLaszip(PointView& dst, const Key& key,
-        const uint64_t nodeId) const
+                               const uint64_t nodeId,
+                               std::unique_ptr<PointTable>& pointTable) const
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: Why did you have to give the pointTable as an input? The previous logic seems to be doing what you need

File: io/EptReader.cpp [hunk 2]
```diff
@@ -630,10 +639,10 @@ uint64_t EptReader::readLaszip(PointView& dst, const Key& key,
     reader.setOptions(options);
 
     std::unique_lock<std::mutex> lock(m_mutex);
-    reader.prepare(table);  // Geotiff SRS initialization is not thread-safe.
+    reader.prepare(*pointTable);  // Geotiff SRS initialization is not thread-safe.
     lock.unlock();
 
-    const auto views(reader.execute(table));
+    const auto views(reader.execute(*pointTable));
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: Not sure why you had to change this fucntion (same for readBinary)

File: io/EptReader.cpp [hunk 3]
```diff
@@ -767,5 +775,89 @@ Dimension::Type EptReader::getCoercedTypeTest(const NL::json& j)
...
+    DimTypeList dims = m_pointView->dimTypes();
+	char* buffer = new char[m_pointView->pointSize()];
+    m_pointView->getPackedPoint(dims, m_currentIndex++, buffer);
+	point.setPackedData(dims, buffer);
+    delete[] buffer;
+}
+
+bool EptReader::processOne(PointRef& point)
+{
+    if (m_keys.size() == 0 && m_nodeId == 1)
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: Could you please add a comment explaining what this does?

File: io/EptReader.cpp [hunk 4]
```diff
@@ -767,5 +775,88 @@ Dimension::Type EptReader::getCoercedTypeTest(const NL::json& j)
...
+    for (const auto& entry : m_overlaps)
+    {
+        const Key& key(entry.first);
+        m_keys.push_back(key);
+    }
+}
+
+void EptReader::fillPoint(PointRef& point)
+{
+    DimTypeList dims = m_pointView->dimTypes();
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: could you not simply set the PointRef to point to the current in the view as oppposed to cpoying the data?

**Review by nicolas-chaulet — CHANGES_REQUESTED** [review]
I think we are going  in the right direction. I am simply wondering if we could amke it a bit simpler, in particular the loadNextOverlap method.

File: io/EptReader.cpp [hunk 5]
```diff
@@ -767,5 +775,88 @@ Dimension::Type EptReader::getCoercedTypeTest(const NL::json& j)
...
+{
+    Key key(m_keys.at(0));
+    m_keys.erase(m_keys.begin());
+    log()->get(LogLevel::Debug)
+        << "Streaming Data " << m_nodeId << "/" << m_overlaps.size() << ": "
+        << key.toString() << std::endl;
+
+    uint64_t startId(0);
+	m_pointTable.reset(new PointTable());
+	if (m_info->dataType() == EptInfo::DataType::Laszip)
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: lasZip and Binary should be treated in the same way (just like in the run method)

File: io/EptReader.cpp [hunk 6]
```diff
@@ -588,9 +587,15 @@ PointViewSet EptReader::run(PointViewPtr view)
             uint64_t startId(0);
 
             if (m_info->dataType() == EptInfo::DataType::Laszip)
-                startId = readLaszip(*view, key, nodeId);
+            {
+				std::unique_ptr<PointTable> table(nullptr);
+                startId = readLaszip(*view, key, nodeId,table);
+                table.reset(nullptr);
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: This extra table should not be necessary

File: io/EptReader.cpp (hunk 1 above)
Comment (pravinshinde825) [reply io/EptReader.cpp]: Done!! 
Reverted to original.

File: io/EptReader.cpp (hunk 3 above)
Comment (pravinshinde825) [reply io/EptReader.cpp]: No need!! Removed this logic.

File: io/EptReader.cpp (hunk 4 above)
Comment (pravinshinde825) [reply io/EptReader.cpp]: `=` operator is deleted from `PointRef`, so we cannot directly set it.
We've to copy.

File: io/EptReader.cpp (hunk 5 above)
Comment (pravinshinde825) [reply io/EptReader.cpp]: Done!!

File: io/EptReader.cpp (hunk 6 above)
Comment (pravinshinde825) [reply io/EptReader.cpp]: Done!! 
Reverted to original.

File: io/EptReader.cpp [hunk 7]
```diff
@@ -588,9 +587,13 @@ PointViewSet EptReader::run(PointViewPtr view)
             uint64_t startId(0);
 
             if (m_info->dataType() == EptInfo::DataType::Laszip)
+            {
                 startId = readLaszip(*view, key, nodeId);
+			}
             else
+            {
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: remove {} as per rhe rest of PDAL syntax

File: io/EptReader.cpp [hunk 8]
```diff
@@ -613,14 +616,14 @@ PointViewSet EptReader::run(PointViewPtr view)
...
-        const uint64_t nodeId) const
+                               const uint64_t nodeId) const
 {
     // If the file is remote (HTTP, S3, Dropbox, etc.), getLocalHandle will
     // download the file and `localPath` will return the location of the
     // downloaded file in a temporary directory.  Otherwise it's a no-op.
     auto handle(m_ep->getLocalHandle("ept-data/" + key.toString() + ".laz"));
 
-    PointTable table;
+	PointTable table;
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: fix indent

File: io/EptReader.cpp [hunk 9]
```diff
@@ -653,8 +656,7 @@ uint64_t EptReader::readLaszip(PointView& dst, const Key& key,
     return startId;
 }
 
-uint64_t EptReader::readBinary(PointView& dst, const Key& key,
-        const uint64_t nodeId) const
+uint64_t EptReader::readBinary(PointView& dst, const Key& key, const uint64_t nodeId) const
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: undo remove break line

File: io/EptReader.cpp [hunk 10]
```diff
@@ -767,5 +769,82 @@ Dimension::Type EptReader::getCoercedTypeTest(const NL::json& j)
...
+{
+    std::map<Key, uint64_t>::iterator itr = m_overlaps.begin();
+    std::advance(itr, m_nodeId-1);
+    Key key(itr->first);
+    log()->get(LogLevel::Debug)
+        << "Streaming Data " << m_nodeId << "/" << m_overlaps.size() << ": "
+        << key.toString() << std::endl;
+
+    uint64_t startId(0);
+	m_pointTable.reset(new PointTable());
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: Fix indentation throughout please

Comment (nicolas-chaulet) [inline io/EptReader.cpp]: Better to write:
```
m_pointTable.reset(new PointTable(m_pointTable->layout()))
```
Then you don't need the for loop below

File: io/EptReader.cpp [hunk 11]
```diff
@@ -767,5 +769,82 @@ Dimension::Type EptReader::getCoercedTypeTest(const NL::json& j)
...
+        readAddon(*m_pointView, key, *addon, startId);
+    }
+    m_nodeId++;
+}
+
+void EptReader::fillPoint(PointRef& point)
+{
+    DimTypeList dims = m_pointView->dimTypes();
+	char* buffer = new char[m_pointView->pointSize()];
+    m_pointView->getPackedPoint(dims, m_currentIndex++, buffer);
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: take the m_m_currentIndex++ out of the function call please

File: io/EptReader.cpp [hunk 12]
```diff
@@ -767,5 +769,82 @@ Dimension::Type EptReader::getCoercedTypeTest(const NL::json& j)
...
+	point.setPackedData(dims, buffer);
+    delete[] buffer;
+}
+
+bool EptReader::processOne(PointRef& point)
+{
+    if (m_nodeId > m_overlaps.size() && m_currentIndex==-1)
+    {
```
Comment (nicolas-chaulet) [inline io/EptReader.cpp]: remove {} as per PDAL general syntax

File: io/EptReader.hpp [hunk 13]
```diff
@@ -71,6 +74,14 @@ class PDAL_DLL EptReader : public Reader
     std::string getName() const override;
 
 private:
+    uint64_t m_nodeId = 1;
```
Comment (nicolas-chaulet) [inline io/EptReader.hpp]: Could you please group those member variables at the bottom of the class definition with the otehr member variables?

File: io/EptReader.hpp [hunk 14]
```diff
@@ -71,6 +74,14 @@ class PDAL_DLL EptReader : public Reader
...
+    std::unique_ptr<PointTable> m_pointTable;
```
Comment (nicolas-chaulet) [inline io/EptReader.hpp]: this point table and view act as buffers, could you please rename to m_bufferPointTable and m_bufferPointView?

**Review by nicolas-chaulet — CHANGES_REQUESTED** [review]
Minor changes, looks a lot better. Could you also please add a test for the streaming interface? You can look at the streaming test in the las reader test as an example. This would be required in order to submit a PR to the upstream. And to our repo as well...

File: io/EptReader.cpp (hunk 10 above)
Comment (pravinshinde825) [reply io/EptReader.cpp]: `PointTable` donot have constructor to accept `PointLayoutPtr` or `PointLayout`.

File: io/EptReader.cpp (hunk 11 above)
Comment (pravinshinde825) [reply io/EptReader.cpp]: Done!

**Comment by pravinshinde825** [comment]
@nicolas-chaulet , I've added test for EPT streamable support and also resolved all comments.

**Comment by nicolas-chaulet** [comment]
I see what you mean, let me rearrange that a bit then. The logic is a bit confusing

**Comment by nicolas-chaulet** [comment]
We don't want warnings in the code just to avoid having an extra variable

**Review by nicolas-chaulet — APPROVED** [review]
Great job @pravinshinde825 !! Feel free to edit some more but it is good to go from my end

**Comment by pravinshinde825** [comment]
Thanks @nicolas-chaulet .

I think I should proceed to create PR on [PDAL/PDAL](https://github.com/PDAL/PDAL) .
What do you think @smileprem ?

**Comment by premkumar-masilamani** [comment]
@pravinshinde825 - Good work. I will merge this PR now. Please proceed with the PR to PDAL.

**Comment by premkumar-masilamani** [comment]
@pravinshinde825 - I see 20+ files changed between `helix-master` and `master` in our fork. Please merge to our `master` first via PR and then create a PR to PDAL.

**Comment by nicolas-chaulet** [comment]
Please could we not do that? I’d like to leave master an exact copy of pdal
master

**Comment by pravinshinde825** [comment]
Just to avoid confusions, I've created new branch `streamable-ept-reader` committed changes related to streamable EPT reader to this branch and created PR to PDAL/PDAL from it: https://github.com/PDAL/PDAL/pull/2603

**Comment by premkumar-masilamani** [comment]
So, the idea is to update the PDAL public repo first from a branch and then update our fork?. I am good with that approach.

**Comment by nicolas-chaulet** [comment]
Yes! Merging PRs into master would make us out of sync with PDAL repo

**Pull request merged**